File: scanners/integrations/webhook_connector.py
"""
Generic Webhook Ticketing Connector (Python) — Phase 14.1

Implements connector interface for generic webhooks with HMAC-SHA256 signature.
Includes all 8 mandatory fields:
- asset ID
- finding ID
- severity
- owner
- evidence link
- remediation
- CBOM reference
- risk score
"""

from datetime import datetime, timezone
import hashlib
import hmac
import json
from typing import Any, Callable, Dict, Optional
import urllib.request
import urllib.error
import uuid

from scanners.integrations.base_connector import BaseTicketingConnector
from scanners.integrations.ticket_request import TicketRequest, TicketResponse


class WebhookConnector(BaseTicketingConnector):
    def __init__(
        self,
        name: str = "webhook-default",
        config: Optional[Dict[str, Any]] = None,
        http_client: Optional[Callable[..., Any]] = None,
    ):
        super().__init__(
            name=name,
            connector_type="webhook",
            config=config,
            http_client=http_client,
        )

    def validate_config(self, config: Dict[str, Any]) -> None:
        super().validate_config(config)
        if not config.get("url"):
            raise ValueError("Webhook connector requires 'url'")

    @property
    def url(self) -> str:
        return str(self.config.get("url", ""))

    @property
    def secret(self) -> str:
        return str(self.config.get("secret", ""))

    @property
    def max_retries(self) -> int:
        return int(self.config.get("max_retries", 2))

    def format_payload(self, ticket_request: TicketRequest) -> Dict[str, Any]:
        return {
            "event": "ecdat.finding.ticket",
            "eventId": f"evt_{uuid.uuid4()}",
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "ticket": {
                "asset_id": ticket_request.asset_id,
                "finding_id": ticket_request.finding_id,
                "severity": ticket_request.severity,
                "owner": ticket_request.owner,
                "evidence_link": ticket_request.evidence_link,
                "remediation": ticket_request.remediation,
                "cbom_ref": ticket_request.cbom_ref,
                "risk_score": ticket_request.risk_score,
                "title": ticket_request.title,
                "description": ticket_request.description,
            },
            "metadata": dict(ticket_request.metadata),
        }

    def compute_signature(self, payload_string: str, timestamp: str) -> Optional[str]:
        if not self.secret:
            return None
        mac = hmac.new(
            self.secret.encode("utf-8"),
            f"{timestamp}.{payload_string}".encode("utf-8"),
            hashlib.sha256,
        )
        return f"sha256={mac.hexdigest()}"

    def send_create_request(
        self, ticket_request: TicketRequest, payload: Dict[str, Any]
    ) -> TicketResponse:
        payload_str = json.dumps(payload)
        timestamp = payload.get("timestamp", datetime.now(timezone.utc).isoformat())

        headers = {
            "Content-Type": "application/json",
            "User-Agent": "ECDAT-Python-Webhook/1.0",
            "X-ECDAT-Event": payload.get("event", "ecdat.finding.ticket"),
            "X-ECDAT-Event-Id": payload.get("eventId", ""),
            "X-ECDAT-Timestamp": timestamp,
        }
        if self.config.get("headers"):
            headers.update(self.config["headers"])

        signature = self.compute_signature(payload_str, timestamp)
        if signature:
            headers["X-ECDAT-Signature"] = signature

        if self.http_client:
            data = self.http_client(self.url, method="POST", json_payload=payload, headers=headers)
        else:
            req = urllib.request.Request(
                self.url,
                data=payload_str.encode("utf-8"),
                headers=headers,
                method="POST",
            )
            try:
                with urllib.request.urlopen(req) as resp:
                    data = json.loads(resp.read().decode("utf-8"))
            except urllib.error.HTTPError as e:
                err_text = e.read().decode("utf-8")
                raise RuntimeError(f"Webhook endpoint returned HTTP {e.code}: {err_text}")

        ticket_id = data.get("ticketId") or data.get("id") or payload.get("eventId")
        ticket_url = data.get("ticketUrl") or data.get("url") or self.url

        return TicketResponse(
            success=True,
            ticket_id=str(ticket_id),
            ticket_url=str(ticket_url),
            connector_type="webhook",
            status=data.get("status", "DISPATCHED"),
            raw_response=data,
        )

File: backend/src/integrations/ticketing/webhook_connector.js
/**
 * Generic Webhook Ticketing Connector — Phase 14.1
 *
 * Implements connector interface for arbitrary enterprise ticketing gateways,
 * SOAR platforms (Splunk SOAR, Palo Alto Cortex XSOAR), and event buses.
 * Includes all 8 mandatory fields:
 * - asset ID
 * - finding ID
 * - severity
 * - owner
 * - evidence link
 * - remediation
 * - CBOM reference
 * - risk score
 */

const crypto = require("crypto");
const { BaseTicketingConnector } = require("./base_connector");
const { TicketResponse } = require("../../domain/contracts");

class WebhookConnector extends BaseTicketingConnector {
  constructor(options = {}) {
    super({
      name: options.name || "webhook-default",
      connectorType: "webhook",
      config: options.config || {},
      fetchFn: options.fetchFn,
    });
  }

  validateConfig(config) {
    super.validateConfig(config);
    if (!config.url) throw new Error("Webhook connector requires 'url'");
  }

  get url() {
    return this.config.url;
  }

  get secret() {
    return this.config.secret || "";
  }

  get maxRetries() {
    return Number.isInteger(this.config.maxRetries) ? this.config.maxRetries : 2;
  }

  /**
   * Pure formatter generating the standardized event payload with all 8 mandatory fields.
   */
  formatPayload(ticketRequest) {
    return {
      event: "ecdat.finding.ticket",
      eventId: `evt_${crypto.randomUUID()}`,
      timestamp: new Date().toISOString(),
      ticket: {
        asset_id: ticketRequest.assetId,
        finding_id: ticketRequest.findingId,
        severity: ticketRequest.severity,
        owner: ticketRequest.owner,
        evidence_link: ticketRequest.evidenceLink,
        remediation: ticketRequest.remediation,
        cbom_ref: ticketRequest.cbomRef,
        risk_score: ticketRequest.riskScore,
        title: ticketRequest.title,
        description: ticketRequest.description,
      },
      metadata: ticketRequest.metadata || {},
    };
  }

  /**
   * Computes HMAC-SHA256 signature for payload verification.
   */
  computeSignature(payloadString, timestamp) {
    if (!this.secret) return null;
    const hmac = crypto.createHmac("sha256", this.secret);
    hmac.update(`${timestamp}.${payloadString}`);
    return `sha256=${hmac.digest("hex")}`;
  }

  async sendCreateRequest(ticketRequest, payload) {
    const payloadString = JSON.stringify(payload);
    const timestamp = payload.timestamp || new Date().toISOString();

    const headers = {
      "Content-Type": "application/json",
      "User-Agent": "ECDAT-Webhook-Connector/1.0",
      "X-ECDAT-Event": payload.event,
      "X-ECDAT-Event-Id": payload.eventId,
      "X-ECDAT-Timestamp": timestamp,
      ...(this.config.headers || {}),
    };

    const signature = this.computeSignature(payloadString, timestamp);
    if (signature) {
      headers["X-ECDAT-Signature"] = signature;
    }

    let lastError = null;
    let res = null;
    let data = null;

    // Retry loop for transient failures (HTTP 429, 502, 503, 504)
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        res = await this.fetchFn(this.url, {
          method: "POST",
          headers,
          body: payloadString,
        });

        data = await res.json().catch(() => ({}));

        if (res.ok) {
          const ticketId = data.ticketId || data.id || data.eventId || payload.eventId;
          const ticketUrl = data.ticketUrl || data.url || this.url;

          return new TicketResponse({
            success: true,
            ticketId: String(ticketId),
            ticketUrl: String(ticketUrl),
            connectorType: "webhook",
            status: data.status || "DISPATCHED",
            rawResponse: data,
          });
        }

        // Only retry on transient server errors
        if (![429, 502, 503, 504].includes(res.status)) {
          throw new Error(`Webhook endpoint returned HTTP ${res.status}: ${JSON.stringify(data)}`);
        }

        lastError = new Error(`Webhook endpoint returned HTTP ${res.status}`);
      } catch (err) {
        lastError = err;
        if (attempt === this.maxRetries) break;
      }
    }

    throw lastError || new Error("Webhook delivery failed after maximum retries");
  }

  async testConnection() {
    const pingPayload = {
      event: "ecdat.webhook.ping",
      eventId: `ping_${crypto.randomUUID()}`,
      timestamp: new Date().toISOString(),
    };
    const payloadStr = JSON.stringify(pingPayload);
    const headers = {
      "Content-Type": "application/json",
      "X-ECDAT-Event": "ecdat.webhook.ping",
      ...(this.config.headers || {}),
    };
    if (this.secret) {
      headers["X-ECDAT-Signature"] = this.computeSignature(payloadStr, pingPayload.timestamp);
    }

    const res = await this.fetchFn(this.url, {
      method: "POST",
      headers,
      body: payloadStr,
    });

    if (!res.ok) {
      throw new Error(`Webhook ping failed with HTTP ${res.status}`);
    }

    return {
      ok: true,
      message: `Successfully reached webhook URL '${this.url}'`,
    };
  }
}

module.exports = {
  WebhookConnector,
};
